internals/cli/cmd_run.go

```diff
@@ -229,21 +229,39 @@ func runDaemon(rcmd *cmdRun, ch chan os.Signal, ready chan<- func()) (err error)
 
 	logger.Debugf("activation done in %v", time.Now().Truncate(time.Millisecond).Sub(t0))
 
+	// The "stop" channel is used by the "enter" command to stop the daemon.
+	var stop chan struct{}
+	if ready != nil {
+		stop = make(chan struct{}, 1)
+	}
+	notifyReady := func() {
+		ready <- func() { close(stop) }
+		close(ready)
+	}
+
 	if !rcmd.Hold {
+		// Start the default services.
 		servopts := client.ServiceOptions{}
 		changeID, err := rcmd.client.AutoStart(&servopts)
 		if err != nil {
 			logger.Noticef("Cannot start default services: %v", err)
-		} else {
+		} else if ready == nil {
```

Comment:
Actually, thinking about this now, I think we can slightly simplify this and fix (half the issue with) #167 so that `pebble run` at least logs errors starting default services. So what about starting the `WaitChange` goroutine in either case so we can log success or failure, but only notifying if `ready != nil`:

```go
// Start the default services (those configured with "startup: enabled").
servopts := client.ServiceOptions{}
changeID, err := rcmd.client.AutoStart(&servopts)
if err != nil {
	logger.Noticef("Cannot start default services: %v", err)
} else {
	// Wait for the default services to actually start and then notify
	// the ready channel (for the "enter" command).
	go func() {
		logger.Debugf("Waiting for default services to autostart with change %s.", changeID)
		_, err := rcmd.client.WaitChange(changeID, nil)
		if err != nil {
			logger.Noticef("Cannot wait for autostart change %s: %v", changeID, err)
		} else {
			logger.Noticef("Started default services with change %s.", changeID)
		}
		if ready != nil {
			notifyReady()
		}
	}()
}
```

Note that addition of a debug log before the `WaitChange` call, as that seems like it'd be useful for debugging.

Reply:
I thought of this before pushing yesterday too. This is nice. I have added https://github.com/canonical/pebble/pull/257/commits/b55bbf8b9b72b167537c886ab2a8a1680ccf6903.